Add annotation processor for generating keep rules for types mentioned only inside generics

Otherwise they will be removed by R8 since it cannot determine whether we are doing reflection on them.

Old issue below:

----

Using com.squareup.retrofit2:retrofit:2.9.0

After upgrading gradle plugin from 

```
com.android.tools.build:gradle:4.1.3
```

to

```
com.android.tools.build:gradle:4.2.2
```

I'm getting following error when running with minifyEnabled true:

```
java.lang.IllegalArgumentException: Method return type must not include a type variable or wildcard
```

Here's how looks the method, which generates the error:

```
    @FormUrlEncoded
    @POST("view_ad") //NON-NLS
    Call<GenericDataResponse<EmptyData>> viewAd(
            @Field("session_id") String sessionId, //NON-NLS
            @Field("id") int id //NON-NLS
    );

    public class EmptyData {

        public EmptyData(){}

    }

    public class GenericDataResponse<T> {

        @SerializedName("result") //NON-NLS
        public String result;

        @SerializedName("data") //NON-NLS
        public T data;

        @SerializedName("mess") //NON-NLS
        public String[] messages;
    }
```

In order to fix this I have to either:
- downgrade to the previous gradle plugin version
- add proguard rule: -keep class com.example.app.EmptyData
- disable proguard/r8

I've read the docs about gradle 4.2 changes and nothing seems to be related to proguard directly. 
Can someone explain what's going on?

@JakeWharton here's a repro:
1. https://github.com/TWiStErRob/repros/tree/master/retrofit/proguard-removed-return-type
2. Run the app (debug variant)

Expected: Activity starts and reads the header:
```

I/okhttp.OkHttpClient: <-- 200 OK https://httpstat.us/200 (556ms)
I/okhttp.OkHttpClient: Content-Length: 31
I/okhttp.OkHttpClient: Content-Type: application/json
I/okhttp.OkHttpClient: Server: Kestrel
I/okhttp.OkHttpClient: Request-Context: appId=cid-v1:1e93d241-20e4-4513-bbd7-f452a16a5d69
I/okhttp.OkHttpClient: Strict-Transport-Security: max-age=2592000
I/okhttp.OkHttpClient: Set-Cookie: ARRAffinity=a9cebb550e254ba9270d015acd372767c78fa4f3f79fc353fb793f2912e673b3;Path=/;HttpOnly;Secure;Domain=httpstat.us
I/okhttp.OkHttpClient: Set-Cookie: ARRAffinitySameSite=a9cebb550e254ba9270d015acd372767c78fa4f3f79fc353fb793f2912e673b3;Path=/;HttpOnly;SameSite=None;Secure;Domain=httpstat.us
I/okhttp.OkHttpClient: Date: Thu, 31 Mar 2022 22:43:38 GMT
I/okhttp.OkHttpClient: {"code":200,"description":"OK"}
I/okhttp.OkHttpClient: <-- END HTTP (31-byte body)
I/System.out: 31
```

Actual:
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.example, PID: 9204
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example/com.example.MainActivity}: java.lang.IllegalArgumentException: Method return type must not include a type variable or wildcard: b.a.i<g.u<?>>
        for method c.a
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3449)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.IllegalArgumentException: Method return type must not include a type variable or wildcard: b.a.i<g.u<?>>
        for method c.a
        at g.z.n(:54)
        at g.z.m(:43)
        at g.w.b(:30)
        at g.v.c(:202)
        at g.v$a.invoke(:160)
        at java.lang.reflect.Proxy.invoke(Proxy.java:1006)
        at $Proxy1.a(Unknown Source)
        at com.example.MainActivity.onCreate(:39)
        at android.app.Activity.performCreate(Activity.java:7994)
        at android.app.Activity.performCreate(Activity.java:7978)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3422)
        	... 11 more
```

Notable points on what makes it work:
 * Change AGP to 7.1.1
 * Change `isMinifyEnabled = false`
 * Uncomment `println(response.body()!!.description)`

<details><summary>That said, adjusting the error message might be useful, because retrace is not helpful here if you have many retrofit interfaces. (click to expand for what I tried)</summary>

### `retrace`-ing the stack 
```
%ANDROID_HOME%\cmdline-tools\latest\bin\retrace.bat build/outputs/mapping/debug/mapping.txt stack.txt
```
gives me:
```
     Caused by: java.lang.IllegalArgumentException: Method return type must not include a type variable or wildcard: b.a.i<g.u<?>>
        for method c.a
        at retrofit2.Utils.methodError(Utils.java:54)
        at retrofit2.Utils.methodError(Utils.java:43)
        at retrofit2.ServiceMethod.parseAnnotations(ServiceMethod.java:30)
        at retrofit2.Retrofit.loadServiceMethod(Retrofit.java:202)
```
which doesn't help, because the most relevant parts are still obfuscated.

### Faking a stack trace with known info

(`b.a.i` is a type, `g.u` is a type, `c` is a type, and `a` is a method, `?` is key, but unknown)
```
        at g.u.helpMe()
        at b.a.i.helpMe()
        at c.a()
        at g.z.n(:54)
```
gives
```
        at retrofit2.Response.helpMe(Response.java)
        at io.reactivex.Single.helpMe(Single.java)
        at c.a()
        at retrofit2.Utils.methodError(Utils.java:54)
```

### Manually looking at mapping file

I searched for `-> .*\.c:` regex (because the package of `c.a` is missing), but this could've yield so many false positives in a normal project.
```
com.example.RetrofitService -> com.example.c:
    io.reactivex.Single makeRequest() -> a
```
so `c.a` == `com.example.RetrofitService.makeRequest()`

</details>

I would propose changing `for method c.a` part of the error message to be faking a StackTraceElement:
```
  at fully.qualified.class.method()
```

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -18,6 +18,8 @@ okhttp = "3.14.9"
 protobuf = "3.25.2"
 robovm = "2.3.14"
 kotlinx-serialization = "1.6.2"
+autoService = "1.1.1"
+incap = "1.0.0"
 
 [libraries]
 androidPlugin = { module = "com.android.tools.build:gradle", version = "8.2.2" }
@@ -40,6 +42,12 @@ animalSnifferAnnotations = { module = "org.codehaus.mojo:animal-sniffer-annotati
 protobufPlugin = "com.google.protobuf:protobuf-gradle-plugin:0.9.4"
 protobuf = { module = "com.google.protobuf:protobuf-java", version.ref = "protobuf" }
 protoc = { module = "com.google.protobuf:protoc", version.ref = "protobuf" }
+
+incap-runtime = { module = "net.ltgt.gradle.incap:incap", version.ref = "incap" }
+incap-processor = { module = "net.ltgt.gradle.incap:incap-processor", version.ref = "incap" }
+
+autoService-annotations = { module = "com.google.auto.service:auto-service-annotations", version.ref = "autoService" }
+autoService-compiler = { module = "com.google.auto.service:auto-service", version.ref = "autoService" }
 
 kotlinCoroutines = { module = "org.jetbrains.kotlinx:kotlinx-coroutines-core", version = "1.7.3" }
 kotlinx-serialization-core = { module = "org.jetbrains.kotlinx:kotlinx-serialization-core", version.ref = "kotlinx-serialization" }
@@ -71,3 +79,4 @@ jsoup = { module = "org.jsoup:jsoup", version = "1.17.2" }
 robovm = { module = "com.mobidevelop.robovm:robovm-rt", version.ref = "robovm" }
 googleJavaFormat = "com.google.googlejavaformat:google-java-format:1.19.2"
 ktlint = "com.pinterest.ktlint:ktlint-cli:1.1.1"
+compileTesting = "com.google.testing.compile:compile-testing:0.21.0"

--- a/retrofit-response-type-keeper/README.md
+++ b/retrofit-response-type-keeper/README.md
@@ -1,0 +1,48 @@
+# Response Type Keeper
+
+Generates keep rules for types mentioned in generic parameter positions of Retrofit service methods.
+
+## Problem
+
+Given a service method like
+```java
+@GET("users/{id}")
+Call<User> getUser(
+  @Path("id") String id);
+```
+
+If you execute this request and do not actually use the returned `User` instance, R8 will remove it
+and replace the return type as `Call<?>`. This fails Retrofit's runtime validation since a wildcard
+is not a valid type to pass to a converter. Note: this removal only occurs if the Retrofit's service
+method definition is the only reference to `User`.
+
+## Solution
+
+This module contains an annotation processor which looks at each Retrofit method and generates
+explicit `-keep` rules for the types mentioned.
+
+Add it to Gradle Java projects with
+```groovy
+annotationProcessor 'com.squareup.retrofit2:response-type-keeper:<version>'
+```
+Or Gradle Kotlin projects with
+```groovy
+kapt 'com.squareup.retrofit2:response-type-keeper:<version>'
+```
+
+For other build systems, the `com.squareup.retrofit2:response-type-keeper` needs added to the Java
+compiler `-processor` classpath.
+
+For the example above, the annotation processor's generated file would contain
+```
+-keep com.example.User
+```
+
+This works for nested generics, such as `Call<ApiResponse<User>>`, which would produce:
+```
+-keep com.example.ApiResponse
+-keep com.example.User
+```
+
+It also works on Kotlin `suspend` functions which turn into a type like
+`Continuation<? extends User>` in the Java bytecode.

--- a/retrofit-response-type-keeper/build.gradle
+++ b/retrofit-response-type-keeper/build.gradle
@@ -1,0 +1,15 @@
+apply plugin: 'org.jetbrains.kotlin.jvm'
+apply plugin: 'org.jetbrains.kotlin.kapt'
+apply plugin: 'com.vanniktech.maven.publish'
+
+dependencies {
+  compileOnly libs.autoService.annotations
+  compileOnly libs.incap.runtime
+  kapt libs.autoService.compiler
+  kapt libs.incap.processor
+
+  testImplementation libs.junit
+  testImplementation libs.compileTesting
+  testImplementation libs.truth
+  testImplementation projects.retrofit
+}

--- a/retrofit-response-type-keeper/gradle.properties
+++ b/retrofit-response-type-keeper/gradle.properties
@@ -1,0 +1,3 @@
+POM_ARTIFACT_ID=response-type-keeper
+POM_NAME=Response Type Keeper
+POM_DESCRIPTION=Annotation processor to generate R8 keep rules for types mentioned in generics.

--- a/retrofit-response-type-keeper/src/main/kotlin/retrofit2/keeper/RetrofitResponseTypeKeepProcessor.kt
+++ b/retrofit-response-type-keeper/src/main/kotlin/retrofit2/keeper/RetrofitResponseTypeKeepProcessor.kt
@@ -1,0 +1,106 @@
+/*
+ * Copyright (C) 2024 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package retrofit2.keeper
+
+import com.google.auto.service.AutoService
+import javax.annotation.processing.AbstractProcessor
+import javax.annotation.processing.Processor
+import javax.annotation.processing.RoundEnvironment
+import javax.lang.model.SourceVersion
+import javax.lang.model.element.ExecutableElement
+import javax.lang.model.element.TypeElement
+import javax.lang.model.type.DeclaredType
+import javax.lang.model.type.TypeMirror
+import javax.lang.model.type.WildcardType
+import javax.tools.StandardLocation.CLASS_OUTPUT
+import net.ltgt.gradle.incap.IncrementalAnnotationProcessor
+import net.ltgt.gradle.incap.IncrementalAnnotationProcessorType.ISOLATING
+
+@AutoService(Processor::class)
+@IncrementalAnnotationProcessor(ISOLATING)
+class RetrofitResponseTypeKeepProcessor : AbstractProcessor() {
+  override fun getSupportedSourceVersion() = SourceVersion.latestSupported()
+  override fun getSupportedAnnotationTypes() = setOf(
+    "retrofit2.http.DELETE",
+    "retrofit2.http.GET",
+    "retrofit2.http.HEAD",
+    "retrofit2.http.HTTP",
+    "retrofit2.http.OPTIONS",
+    "retrofit2.http.PATCH",
+    "retrofit2.http.POST",
+    "retrofit2.http.PUT",
+  )
+
+  override fun process(
+    annotations: Set<TypeElement>,
+    roundEnv: RoundEnvironment,
+  ): Boolean {
+    val elements = processingEnv.elementUtils
+    val types = processingEnv.typeUtils
+
+    val methods = supportedAnnotationTypes
+      .mapNotNull(elements::getTypeElement)
+      .flatMap(roundEnv::getElementsAnnotatedWith)
+
+    val elementToReferencedTypes = mutableMapOf<TypeElement, MutableSet<String>>()
+    for (method in methods) {
+      val executableElement = method as ExecutableElement
+
+      val serviceType = method.enclosingElement as TypeElement
+      val referenced = elementToReferencedTypes.getOrPut(serviceType, ::LinkedHashSet)
+
+      val returnType = executableElement.returnType as DeclaredType
+      returnType.recursiveParameterizedTypesTo(referenced)
+
+      // Retrofit has special support for 'suspend fun' in Kotlin which manifests as a
+      // final Continuation parameter whose generic type is the declared return type.
+      executableElement.parameters
+        .lastOrNull()
+        ?.asType()
+        ?.takeIf { types.erasure(it).toString() == "kotlin.coroutines.Continuation" }
+        ?.let { (it as DeclaredType).typeArguments.single() }
+        ?.recursiveParameterizedTypesTo(referenced)
+    }
+
+    for ((element, referencedTypes) in elementToReferencedTypes) {
+      val typeName = element.qualifiedName.toString()
+      val outputFile = "META-INF/proguard/retrofit-response-type-keeper-$typeName.pro"
+      val rules = processingEnv.filer.createResource(CLASS_OUTPUT, "", outputFile, element)
+      rules.openWriter().buffered().use { w ->
+        w.write("# $typeName\n")
+        for (referencedType in referencedTypes.sorted()) {
+          w.write("-keep,allowobfuscation,allowoptimization class $referencedType\n")
+        }
+      }
+    }
+    return false
+  }
+
+  private fun TypeMirror.recursiveParameterizedTypesTo(types: MutableSet<String>) {
+    when (this) {
+      is WildcardType -> {
+        extendsBound?.recursiveParameterizedTypesTo(types)
+        superBound?.recursiveParameterizedTypesTo(types)
+      }
+      is DeclaredType -> {
+        for (typeArgument in typeArguments) {
+          typeArgument.recursiveParameterizedTypesTo(types)
+        }
+        types += (asElement() as TypeElement).qualifiedName.toString()
+      }
+    }
+  }
+}

--- a/retrofit-response-type-keeper/src/test/kotlin/retrofit2/keeper/RetrofitResponseTypeKeepProcessorTest.kt
+++ b/retrofit-response-type-keeper/src/test/kotlin/retrofit2/keeper/RetrofitResponseTypeKeepProcessorTest.kt
@@ -1,0 +1,162 @@
+/*
+ * Copyright (C) 2024 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package retrofit2.keeper
+
+import com.google.common.truth.Truth.assertAbout
+import com.google.testing.compile.JavaFileObjects
+import com.google.testing.compile.JavaSourceSubjectFactory.javaSource
+import java.nio.charset.StandardCharsets.UTF_8
+import javax.tools.StandardLocation.CLASS_OUTPUT
+import org.junit.Test
+
+class RetrofitResponseTypeKeepProcessorTest {
+  @Test
+  fun allHttpMethods() {
+    val service = JavaFileObjects.forSourceString(
+      "test.Service",
+      """
+      package test;
+      import retrofit2.*;
+      import retrofit2.http.*;
+
+      class DeleteUser {}
+      class GetUser {}
+      class HeadUser {}
+      class HttpUser {}
+      class OptionsUser {}
+      class PatchUser {}
+      class PostUser {}
+      class PutUser {}
+
+      interface Service {
+        @DELETE("/") Call<DeleteUser> delete();
+        @GET("/") Call<GetUser> get();
+        @HEAD("/") Call<HeadUser> head();
+        @HTTP(method = "CUSTOM", path = "/") Call<HttpUser> http();
+        @OPTIONS("/") Call<OptionsUser> options();
+        @PATCH("/") Call<PatchUser> patch();
+        @POST("/") Call<PostUser> post();
+        @PUT("/") Call<PutUser> put();
+      }
+    """.trimIndent(),
+    )
+
+    assertAbout(javaSource())
+      .that(service)
+      .processedWith(RetrofitResponseTypeKeepProcessor())
+      .compilesWithoutError()
+      .and()
+      .generatesFileNamed(
+        CLASS_OUTPUT,
+        "",
+        "META-INF/proguard/retrofit-response-type-keeper-test.Service.pro",
+      ).withStringContents(
+        UTF_8,
+        """
+        |# test.Service
+        |-keep,allowobfuscation,allowoptimization class retrofit2.Call
+        |-keep,allowobfuscation,allowoptimization class test.DeleteUser
+        |-keep,allowobfuscation,allowoptimization class test.GetUser
+        |-keep,allowobfuscation,allowoptimization class test.HeadUser
+        |-keep,allowobfuscation,allowoptimization class test.HttpUser
+        |-keep,allowobfuscation,allowoptimization class test.OptionsUser
+        |-keep,allowobfuscation,allowoptimization class test.PatchUser
+        |-keep,allowobfuscation,allowoptimization class test.PostUser
+        |-keep,allowobfuscation,allowoptimization class test.PutUser
+        |
+        """.trimMargin(),
+      )
+  }
+
+  @Test
+  fun nesting() {
+    val service = JavaFileObjects.forSourceString(
+      "test.Service",
+      """
+      package test;
+      import retrofit2.*;
+      import retrofit2.http.*;
+
+      class One<T> {}
+      class Two<T> {}
+      class Three {}
+
+      interface Service {
+        @GET("/") Call<One<Two<Three>>> get();
+      }
+    """.trimIndent(),
+    )
+
+    assertAbout(javaSource())
+      .that(service)
+      .processedWith(RetrofitResponseTypeKeepProcessor())
+      .compilesWithoutError()
+      .and()
+      .generatesFileNamed(
+        CLASS_OUTPUT,
+        "",
+        "META-INF/proguard/retrofit-response-type-keeper-test.Service.pro",
+      ).withStringContents(
+        UTF_8,
+        """
+        |# test.Service
+        |-keep,allowobfuscation,allowoptimization class retrofit2.Call
+        |-keep,allowobfuscation,allowoptimization class test.One
+        |-keep,allowobfuscation,allowoptimization class test.Three
+        |-keep,allowobfuscation,allowoptimization class test.Two
+        |
+        """.trimMargin(),
+      )
+  }
+
+  @Test
+  fun kotlinSuspend() {
+    val service = JavaFileObjects.forSourceString(
+      "test.Service",
+      """
+      package test;
+      import kotlin.coroutines.Continuation;
+      import retrofit2.*;
+      import retrofit2.http.*;
+
+      class Body {}
+
+      interface Service {
+        @GET("/") Object get(Continuation<? extends Body> c);
+      }
+    """.trimIndent(),
+    )
+
+    assertAbout(javaSource())
+      .that(service)
+      .processedWith(RetrofitResponseTypeKeepProcessor())
+      .compilesWithoutError()
+      .and()
+      .generatesFileNamed(
+        CLASS_OUTPUT,
+        "",
+        "META-INF/proguard/retrofit-response-type-keeper-test.Service.pro",
+      ).withStringContents(
+        UTF_8,
+        """
+        |# test.Service
+        |-keep,allowobfuscation,allowoptimization class java.lang.Object
+        |-keep,allowobfuscation,allowoptimization class test.Body
+        |
+        """.trimMargin(),
+      )
+  }
+}

--- a/settings.gradle
+++ b/settings.gradle
@@ -15,6 +15,8 @@ include ':retrofit:test-helpers'
 
 include ':retrofit-mock'
 
+include ':retrofit-response-type-keeper'
+
 include ':retrofit-adapters:guava'
 include ':retrofit-adapters:java8'
 include ':retrofit-adapters:rxjava'

